### ISSUE TITLE
Metric `overlay.delay.write-queue` has negative values (Version: 12.5.0-32)

### ISSUE THREAD

**Issue by jacekn** [stellar/stellar-core#2518]
## Issue Description

I have stellar-core 12.5.0-32 that exposes negative values of the `overlay.delay.write-queue` metric (for example sum or mean):
```
,"overlay.delay.write-queue":{
"type":"timer",
"count":12216602352,
"event_type":"calls",
"rate_unit":"s",
"mean_rate":4504.95,
"1_min_rate":6277.53,
"5_min_rate":6355.12,
"15_min_rate":6257.9,
"duration_unit":"ms",
"min":0,
"max":149469,
"mean":-633.807,
"stddev":7763.77,
"sum":-7.74296e+12,
"median":0,
"75%":0.001,
"95%":5.3996,
"98%":2962.81,
"99%":14027.3,
"99.9%":66916.1
}
```

## Steps to Reproduce

On core 12.5.0-32: `curl 127.0.0.1:11626/metrics` and look at metrics

## Expected Result

Timers have positive values

## Actual Result

`overlay.delay.write-queue` has negative sum and mean. In my case this is the only affected metric.

## Your Environment and Setup

* stellar-core 12.5.0-32
* Ubuntu 16.04.6 LTS
* installed using apt.stellar.org deb package

## Additional context

I noticed this because the [stellar-core-prometheus-exporter](https://github.com/stellar/stellar-core-prometheus-exporter) stops returning data when it sees negative values for sum.

**Comment by MonsieurNicolas** [comment]
Weird, looks `TimestampedMessage::recordWriteTiming` is called when it shouldn't or we overflow duration somehow on the target system?

**Comment by MonsieurNicolas** [comment]
@graydon of note, I think that the change to steady clock may not be enough for this particular one: we suspect overflow in this case as we accumulate timing for all writes (look at the value of `count: 12216602352`).

Basically, writes performed take in the order of a second, which in nano is 1e9, and as we have many writes done, we end up with `12216602352*1e9 ~= 1.22e19 > 2^63`.

The fix might be to either
* change unit for that metric (or class of metrics). not sure how that works: medida seems to be hardcoded to use nanoseconds.
* change medida to automatically reset metrics that overflow. Downstream systems like prometheus actually detect when this happens and compensate.

**Comment by graydon** [comment]
I looked at this a bit today and agree with your diagnosis: the above peer is recording 4000-ish timer-events per second. One timer-event per write on a large-ish number of peers -- that's not impossibly-fast. That means that for it to hit 12216602352 events it has to have been running for hmm .. 35 days without a reboot (yay!) which is surprising but not impossible.

And yeah at nanosecond resolution that'll overflow int64_t in the sum, and then the mean is derived from the sum so ... gross.

However: the event-count isn't overflowing (and won't -- at 4000 events/sec it'll take 10**38 years to overflow). Only sums (and possibly other values derived values) that store or work with cumulative _quantities of nanoseconds_ are. Because nanoseconds throw that 10^9 factor in to begin with. You don't have to accumulate a lot of nanosecond-counts before you get big numbers.

Looking at medida, what's interesting is that the _interfaces_ for dealing with such cumulative nanosecond-counts are all expressed in terms of doubles anyways. Like the `SummarizableInterface` in medida (which histogram implements) looks like so:

~~~
class SummarizableInterface {                                                                                                                                
public:                                                                                                                                                      
  virtual ~SummarizableInterface() {};                                                                                                                       
  virtual double max() const = 0;                                                                                                                            
  virtual double min() const = 0;                                                                                                                            
  virtual double mean() const = 0;                                                                                                                           
  virtual double std_dev() const = 0;                                                                                                                        
  virtual double sum() const = 0;                                                                                                                            
};                                                                                                                                                           
~~~

So this raises the question: why is medida accumulating something in a int64_t when it's going to report it in terms of a double? It appears (working backwards to infer the author's intent here) that it's because the author wanted to store the values like this:

~~~
  std::atomic<std::int64_t> min_;
  std::atomic<std::int64_t> max_;
  std::atomic<std::int64_t> sum_;
  std::atomic<std::uint64_t> count_;
~~~

That is, specifically, I think the author just wanted to use `std::atomic` and it turns out that `std::atomic` isn't supported for floating point types (at least not until C++20), only integral types.

This is readily solvable! We can just switch the accumulation values to double (since that's what the clients want anyways) and protect concurrent access with a `std::mutex`. There's actually already a mutex _in_ that class for protecting the variance calculation. And most measurements I've seen (I'll do some more myself to be sure) suggest that lightly-contended mutexes, in a modern implementation, are probably as fast or faster than a std::atomic anyways.

I've done a sketch of the change that seems to work, I'll do a little more testing and cleanup and post a patch.

**Comment by graydon** [comment]
theoretically fixed by  #2583
